### PR TITLE
Enable sample pool groups in proteomics measurements

### PR DESCRIPTION
**Pull request by sven1103**

### CODE REVIEW

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/measurements/MeasurementDetailsComponent.java [hunk 1]
```diff
@@ -176,23 +188,38 @@ private void createProteomicsGrid() {
...
-    proteomicsMeasurementGrid.addColumn(
-            proteomicsMeasurement -> proteomicsMeasurement.instrument().getName())
-        .setHeader("Name");
-    proteomicsMeasurementGrid.addThemeVariants(GridVariant.LUMO_WRAP_CELL_CONTENT);
+            proteomicsMeasurement.measuredSamples())).setHeader("Sample Codes").setAutoWidth(true);
+    proteomicsMeasurementGrid.addComponentColumn(
+            proteomicsMeasurement -> renderOrganisation().createComponent(proteomicsMeasurement.organisation()))
+        .setHeader("Organisation").setTooltipGenerator(proteomicsMeasurement -> proteomicsMeasurement.organisation().label()).setAutoWidth(true).setFlexGrow(0);
+    proteomicsMeasurementGrid.addColumn(ProteomicsMeasurement::facility).setHeader("Facility").setTooltipGenerator(ProteomicsMeasurement::facility).setAutoWidth(true);
+    proteomicsMeasurementGrid.addComponentColumn(
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/project/measurements/MeasurementDetailsComponent.java:202]: oh I did not know about the ComponentColumn

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/measurements/MeasurementMetadataUploadDialog.java [hunk 2]
```diff
@@ -120,6 +119,100 @@ public MeasurementMetadataUploadDialog(ValidationService validationService,
...
+
+    return new MetadataContent(content.isEmpty() ? null : content.get(0),
+        content.size() > 1 ? content.subList(1, content.size()) : new ArrayList<>());
+  }
+
+  private static boolean isRowNotEmpty(String row) {
+    return row.split("\t").length > 0;
+  }
+
+  private static Result<ProteomicsMeasurementMetadata, String> generatePxPRequest(
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/project/measurements/MeasurementMetadataUploadDialog.java:147]: please adapt the method name. There is no request anymore

**Review by KochTobi — CHANGES_REQUESTED** [review]
Some minor things and some things where the reasoning is missing.

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementMetadata.java [hunk 3]
```diff
@@ -1,10 +1,23 @@
...
  */
 public interface MeasurementMetadata {
 
+  default Optional<String> assignedSamplePoolGroup() {
+    return Optional.empty();
+  }
+
+  default List<SampleCode> associatedSamples() {
+    return Collections.emptyList();
+  }
```
[the remark this answers is not in the document]
Comment (sven1103) [reply project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementMetadata.java:21]: These are properties that all Measurement Metadata will share

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/MeasurementCode.java [hunk 4]
```diff
@@ -97,6 +91,18 @@ public boolean isIMGDomain() {
   public enum MEASUREMENT_PREFIX {
     NGS, MS, IMG
   }
+  static class MeasurementCodeConverter implements AttributeConverter<MeasurementCode, String> {
```
[the remark this answers is not in the document]
Comment (sven1103) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/MeasurementCode.java:94]: what does it do?

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java [hunk 5]
```diff
@@ -38,7 +32,7 @@ public class NGSMeasurement implements MeasurementMetadata {
   @Embedded
   private OntologyTerm instrument;
 
-  @Embedded
+  @Convert(converter = MeasurementCode.MeasurementCodeConverter.class)
```
[the remark this answers is not in the document]
Comment (sven1103) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java:35]: I argue that my explicit notation is easier to navigate to the assigned converter

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/ProteomicsMeasurement.java [hunk 6]
```diff
@@ -129,4 +205,61 @@ public OntologyTerm instrument() {
...
+  public String lcColumn() {
+    return lcColumn;
+  }
+
+  public String lcmsMethod() {
+    return lcmsMethod;
+  }
+
+  public Optional<String> comment() {
+    return Optional.ofNullable(comment.isBlank() ? null : comment);
```
[the remark this answers is not in the document]
Comment (sven1103) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/ProteomicsMeasurement.java:238]: Optional of null value returns Optional.empty. It is just a concise expression

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/ProteomicsMeasurement.java [hunk 7]
```diff
@@ -51,14 +79,55 @@ protected ProteomicsMeasurement() {
...
+  }
+
+  private static void evaluateMandatorMetadata(ProteomicsMethodMetadata method)
+      throws IllegalArgumentException {
+    if (method.instrument() == null) {
+      throw new IllegalArgumentException("Instrument: Missing metadata.");
+    }
+    if (method.facility().isBlank()) {
+      throw new IllegalArgumentException("Facility: Missing metadata");
+    }
```
[the remark this answers is not in the document]
Comment (sven1103) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/ProteomicsMeasurement.java:108]: null if is an Object that is not assigned, else for String values

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/MeasurementCode.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (KochTobi) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/MeasurementCode.java:94]: the same as `@Converter(...)` on the fields in JPA classes. But it is discovered automatically and used as a default for fields of type `MeasurementCode`

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (KochTobi) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java:35]: But you need to make sure you add it everywhere. I'd argue that finding implementations of interfaces is no difficult task nowadays.

[the remark this answers is not in the document]
Comment (sven1103) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java:35]: Why everywhere? This is the only class that represents the JPA entity

**Review by KochTobi — APPROVED** [review]
LGTM

**Pull request merged**